wrappers/wrapper.rb

```diff
@@ -742,89 +748,69 @@ def detect_unfeasible_services(vrp)
       unfeasible
     end
 
-    def check_distances(vrp, unfeasible)
-      unfeasible = check(vrp, :time, unfeasible)
-      unfeasible = check(vrp, :distance, unfeasible)
-      unfeasible = check(vrp, :value, unfeasible)
-
-      # check distances from vehicle depot is feasible
-      vrp.services.each{ |service|
-        # Check via vehicle time-windows
-        reachable_by_a_vehicle = (service.activity ? [service.activity] : service.activities).any?{ |activity|
-          index = activity.point.matrix_index
-          vrp.vehicles.find{ |vehicle|
-            feasible = true
-
-            start_index = vehicle.start_point&.matrix_index
-            end_index = vehicle.end_point&.matrix_index
-            matrix = vrp.matrices.find{ |mat| mat.id == vehicle.matrix_id } if start_index || end_index
+    def service_reachable_by_vehicle_within_timewindows(vrp, activity, vehicle)
+      service_index = activity.point.matrix_index
+      v_start_index = vehicle.start_point&.matrix_index
+      v_end_index = vehicle.end_point&.matrix_index
+      matrix = vrp.matrices.find{ |mat| mat.id == vehicle.matrix_id }
 
-            if vehicle.cost_time_multiplier&.positive? &&
-               !vehicle.cost_late_multiplier&.positive? &&
-               ((vehicle.timewindow&.start && vehicle.timewindow&.end) || vehicle.sequence_timewindows&.size&.positive?)
-              vehicle_available_time = vehicle.timewindow.end - vehicle.timewindow.start if vehicle.timewindow
-              vehicle_available_time = vehicle.sequence_timewindows.collect{ |tw| tw.end - tw.start }.max if vehicle.sequence_timewindows&.size&.positive?
+      vehicle_start = vehicle.timewindow&.start || vehicle.sequence_timewindows.collect(&:start).min || 0
+      vehicle_end = vehicle.timewindow&.end || vehicle.sequence_timewindows.collect(&:end).max || 2**56
+      vehicle_available_time = vehicle_end - vehicle_start
 
-              min_time = (start_index && matrix[:time][start_index][index]).to_f + (end_index && matrix[:time][index][end_index]).to_f + activity.duration
-
-              feasible &&= (vehicle_available_time >= min_time)
-            end
+      earliest_arrival = vehicle_start
+      earliest_arrival += matrix.time[v_start_index][service_index] if v_start_index
+      if !activity.late_multiplier&.positive? && activity.timewindows.any?(&:end)
+        return false if activity.timewindows.all?{ |tw| tw.end && earliest_arrival > tw.end }
+      end
 
-            if feasible &&
-               (start_index || end_index) &&
-               vehicle.cost_distance_multiplier&.positive? &&
-               vehicle.distance
-              min_dist = (start_index && matrix[:distance][start_index][index]).to_f + (end_index && matrix[:distance][index][end_index]).to_f
+      if vehicle_available_time && vehicle.cost_time_multiplier&.positive? && !vehicle.cost_late_multiplier&.positive?
```

Comment:
`vehicle_available_time` always exists what we want is `vehicle_end != 2**56`, right ?

Reply:
If the next check was in fact unnecessary, and if we are going to delete it then we can 
let `vehicle_end = vehicle.timewindow&.end || vehicle.sequence_timewindows.collect(&:end).max`
and do the check with `vehicle_end.nil?` here
and generate the `vehicle_available_time = vehicle_end - vehicle_start` inside this if condition 
instead of calculating `2**56` multiple times

Reply:
Yes I will do something like that. We still need || 2**56 because of next check. 

Instead of checking vehicle_available_time I will just check vehicle_end < 2**56